### DISCUSSION TITLE
Assign default literal to a deconstruction

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#583] [General]
@alrz commented on [Mon May 15 2017](https://github.com/dotnet/roslyn/issues/19513)

**Version Used**: master

**Steps to Reproduce**:

```cs
class C {
    void M(int i , int j) {
        (i, j) = default;
    }
}
```
**Expected Behavior**: compile with no error.

**Actual Behavior**: error CS8131: Deconstruct assignment requires an expression with a type on the right-hand-side.

---

@alrz commented on [Mon May 15 2017](https://github.com/dotnet/roslyn/issues/19513#issuecomment-301584539)

@jcouv Can this be special cased? [would come in handy sometime](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs,07114598e8759e4e).

**Comment by jcouv** [comment]
I had not thought about that. Thanks for the pointer to actual usage of this (copied that snippet below):

```C#
(constant, expression) =
                        IsConstant(expression1) ? (expression1, expression2) :
                        IsConstant(expression2) ? (expression2, expression1) :
                        default((TExpressionSyntax, TExpressionSyntax));
```

The [current feature](https://github.com/dotnet/roslyn/blob/master/docs/features/deconstruction.md) only supports either (1) a tuple literal (which will get some type inference), (2) a tuple type (which is trivial to deconstruct), or (3) some type with `Deconstruct`.
Adding a fourth case for `default` doesn't seem too hard, but we should discuss with LDM.

FYI @gafter

**Comment by svick** [comment]
@alrz 

That specific case already seems to work. [This code compiles fine on Roslyn master:](https://tryroslyn.azurewebsites.net/#b:master/f:r/IYIwzgLgTsDGEAJYBthjAgwgHgCoFEAPABygFN0BLAewDsBlAT1omEID4BYAKAG8eEghKWoQy8MgBMEIatWQJcURgBEyYqAFtKtMpjqRgLABQCh5giXJU6TFmwRkrFMDVoBGADRnzgy6Rc3O1ZCR2cbWgAmb25fIWoAV0R/a1dbZhCkA1YWGLjBROSiAIjghycStNoASh8Eflj841hsowhPMMq3aoQAXjr8hABJMH1aQxMK1Ld3HoB+BGMpwLovTum6SJ6ALgH8kbGJiCXwqq2EBZOuzY7liNmEXcbBwUkyADNgBOQIAG4ePaCcgQBJQWhZcY5RAAQl6CFo32Q/2eCAAvoDfDwAJAiMQSaSgSAweAyOQKA6tEwpFYMDLlU7dZGooA===)

```c#
(constant, expression) =
    IsConstant(expression1) ? (expression1, expression2) :
    IsConstant(expression2) ? (expression2, expression1) :
    default;
```

**Comment by alrz** [comment]
@svick 

Right. A ternary does have a type, whence the `default` is inferred, which in turn, satisfies the deconstruction requirement.

**Comment by alrz** [comment]
Still I think  `(x , y) = default;` could be useful when you want to zero-out some variables, potentially out params, **that are not of the same type** (otherwise could be accomplished via chained assignments).

**Comment by artemious7** [comment]
Perhaps, it would be better to zero-out some variables like this: `x , y = default;` instead of `(x , y) = default;`. Otherwise, it would be confusing, because to me it means deconstructing from a Tuple, which is not really involved here. It should be merely setting default values to each of the existing variables.
